Getting Started redo, Issue 277. (#395)
k8ssandra/k8ssandra: docs/content/en/docs/getting-started/_index.md

---
title: "K8ssandra quick start"
linkTitle: "Quick start"
weight: 1
description: |
  Kick the tires and take K8ssandra for a spin!
---

> Welcome to K8ssandra! This guide gets you up and running with a single-node Apache Cassandra&reg; cluster on Kubernetes. If you're interested in a more detailed component walkthroughs check out the [tasks]({{< ref "topics">}}) section.

Author: I'll fix this at some point, but for now it's transparent to customers.

> Welcome to K8ssandra! This guide gets you up and running with a single-node Apache Cassandra&reg; cluster on Kubernetes. If you're interested in a more detailed component walkthroughs check out the [tasks]({{< ref "topics">}}) section.

Reviewer: When we go to fix it we need to make sure to set up redirects for the old URL

* [Verifying K8ssandra functionality]({{< relref "#verify-your-k8ssandra-installation" >}}): Making sure K8ssandra is working as expected.
* [Retrieve K8ssandra superuser credentials]({{< relref "#superuser" >}}): Getting the K8ssandra superuser name and password so you can access common utilities as well as the Stargate API.

Once these basic configuration and verification steps are completed, you can choose more detailed paths for either a developer or a site reliability engineer.

## Prerequisites

In your local environment the following tools are required for provisioning a K8ssandra cluster:

* [Helm v3+](https://helm.sh/docs/intro/install/)
* [Kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)

As K8ssandra deploys on a K8s cluster, one must be available to target for installation. The K8s environment may be a local version running on your development machine, an on-premises self-hosted environment, or a managed cloud offering.

K8ssandra works with the following versions of Kubernetes either standalone or via a cloud provider:

* 1.16
* 1.17
* 1.18
* 1.19
* 1.20

To verify your K8s version:

```bash

> kubectl version

Author: I saw that but didn't realize it was implemented yet. I'll redo all the examples.

k8ssandra-reaper-k8ssandra-5b6cc959b7-wt22j           1/1     Running     0          3m20s
k8ssandra-reaper-k8ssandra-schema-dnrpk               0/1     Completed   0          3m35s
k8ssandra-reaper-operator-cc46fd5f4-lzq96             1/1     Running     0          7m
prometheus-k8ssandra-kube-prometheus-prometheus-0     2/2     Running     1          6m32s
```

The K8ssandra pods in the example above have the identifier `k8ssandra` either prefixed or inline, since that's the name that was specified when the cluster was created using Helm. If you choose a different cluster name during installation, your pod names will be different.

The actual Cassandra node name from the listing above is `k8ssandra-dc1-default-sts-0` which we'll use throughout the following sections.

Verify the following:

* The K8ssandra pod running Cassandra, `k8ssandra-dc1-default-sts-0` in the example above should show `2/2` as `Ready`.
* The Stargate pod, `k8ssandra-dc1-stargate-6f7f5d6fd6-2dz8f` in the example above should show `1/1` as `Ready`. 

{{% alert title="Important" color="warning" %}}
* The Stargate pod will not show `Ready` until at least 4 minutes have elapsed. 

> * The pod `k8ssandra-reaper-k8ssandra-schema-xxxxx` runs once and does not persist.

Author: You got it!

    k8ssandra-reaper-k8ssandra-reaper-service   ClusterIP   10.104.46.103    <none>        8080/TCP                                                47m
    k8ssandra-seed-service                      ClusterIP   None             <none>        <none>                                                  47m
    kubernetes                                  ClusterIP   10.96.0.1        <none>        443/TCP                                                 53m
    prometheus-operated                         ClusterIP   None             <none>        9090/TCP                                                47m
    ```

1. Verify the name of the Cassandra datacenter:

    ```bash
    kubectl get cassandradatacenters
    NAME   AGE
    dc1    51m
    ```

## Retrieve K8ssandra superuser credentials {#superuser}

You'll need the K8ssandra superuser name and password in order to access Cassandra utilities and do things like generate a Stargate access token.

To retrieve K8ssandra superuser credentials:

1. Retrieve the K8ssandra superuser name:

    ```bash

>     kubectl get secret k8ssandra-superuser -o jsonpath="{.data.username}" | base64 --decode | more

Reviewer 2: Cleaner: `kubectl get secret k8ssandra-superuser -o jsonpath="{.data.username}" | base64 --decode ; echo`

The `; echo` gives you the newline I think you're trying to get.

Author: You are AWESOME! For some reason, in my iterm, it adds a % control character at the end.

    k8ssandra-superuser
    (END)
    ```

1. Retrieve the K8ssandra superuser password:

    ```bash

>     kubectl get secret k8ssandra-superuser -o jsonpath="{.data.password}" | base64 --decode | more

Reviewer 2: see previous comment

* [Helm v3+](https://helm.sh/docs/intro/install/)
* [Kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)

> As K8ssandra deploys on a Kubernetes cluster, one must be available to target for installation. The K8 environment may be a local version running on your development machine, an on-premises self-hosted environment, or a managed cloud offering. To that end the cluster must be up and available to your `kubectl` command:

Author: I can never remember. Fixed.

```bash
minikube start --cpus=4 --memory='8128m' --kubernetes-version=1.18.16 -p k8ssandra
😄  [k8ssandra] minikube v1.17.1 on Darwin 11.2.1
✨  Automatically selected the docker driver. Other choices: hyperkit, ssh
👍  Starting control plane node k8ssandra in cluster k8ssandra
🔥  Creating docker container (CPUs=4, Memory=8128MB) ...
🐳  Preparing Kubernetes v1.18.16 on Docker 20.10.2 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔎  Verifying Kubernetes components...
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "k8ssandra" cluster and "default" namespace by default
```

## Configure the K8ssandra Helm repository

K8ssandra is delivered via a collection of Helm charts for easy installation, so once you've got a suitable K8s environment configured, you'll need to add the K8ssandra Helm chart repositories.

To add the K8ssandra helm chart repos:

1. Add the main K8ssandra Helm chart repo:

    ```bash
    helm repo add k8ssandra https://helm.k8ssandra.io/
    ```

> 2. If you want to access K8ssandra services from outside of the Kubernetes cluster, also add the Traefik Ingress repo (highly recommended):

Reviewer: Do we want to include Traefik here? Depending on your use case it may not be required.

Reviewer: (Read: Should this be moved to the DevOps section and leave this part of the path?)

Author: I'm sort of ambivalent on this. Honestly I don't see why we don't just glom the K8ssandra and Traefik repos into one. It's not like having the traefik repos will harm anything or get in the way.

    ```bash
    helm repo add traefik https://helm.traefik.io/traefik
    ```

3. Finally, update your helm repository listing:

    ```bash
    helm repo update
    ```

{{% alert title="Tip" color="success" %}}
Alternatively, you can download the individual charts directly from the project's [releases](https://github.com/k8ssandra/k8ssandra/releases) page.
{{% /alert %}}

## Install K8ssandra

The K8ssandra helm charts make installation a snap. You can override chart configurations during installation as necessary if you're an advanced user, or make changes after a default installation using `helm upgrade` at a later time.

{{% alert title="Important" color="warning" %}}

> K8ssandra comes out of the box with a set of default values tailored to getting up and running quickly.  Those defaults are intended to be a great starting point for smaller-scale local development but are **not** intended for production deployments.

Reviewer: Add a link to `values.yaml` for the K8ssandra chart. Another option would be a link to the reference page for the chart.

```suggestion
K8ssandra comes out of the box with a set of [default values](https://github.com/k8ssandra/k8ssandra/blob/main/charts/k8ssandra/values.yaml) tailored to getting up and running quickly.  Those defaults are intended to be a great starting point for smaller-scale local development but are **not** intended for production deployments.
```

or

```suggestion
K8ssandra comes out of the box with a set of [default values]({{< ref "/docs/reference/k8ssandra/" >}}) tailored to getting up and running quickly.  Those defaults are intended to be a great starting point for smaller-scale local development but are **not** intended for production deployments.
```

{{% /alert %}}

To install a single node K8ssandra cluster:

1. Copy the following YAML to a file named `k8ssandra.yaml`:

>     ```yaml

Reviewer: Consider replacing with the following YAML:

```yaml
cassandra:
  version: "3.11.10"
  cassandraLibDirVolume:
    storageClass: local-path
    size: 5Gi
  heap:
   size: 1G
   newGenSize: 1G
  resources:
    requests:
      cpu: 1000m
      memory: 2Gi
    limits:
      cpu: 1000m
      memory: 2Gi
  datacenters:
  - name: dc1
    size: 1
    racks:
    - name: default
stargate:
  enabled: true
  replicas: 1
  heapMB: 256
  cpuReqMillicores: 200
  cpuLimMillicores: 1000
```

follow by a rundown of what we're configuring. In this case:

* 1x C* node
  * 1 GB heap
  * 2 GB RAM for container
  * 1 CPU core
  * 5GB storage
  * local-path storage class - NOTE THIS MUST BE ADJUSTED FOR YOUR ENVIRONMENT.
* 1x Stargate node
*

    kube-prometheus-stack:
      grafana:
        adminUser: admin
        adminPassword: admin123
    ```

    The configuration file creates a K8ssandra cluster named `k8ssandra` with a datacenter, `dc1` containing a single Cassandra node.

2. Use `helm install` to install K8ssandra, referring to the example configuration file:

    ```bash
    helm install -f k8ssandra.yaml k8ssandra k8ssandra/k8ssandra
    NAME: k8ssandra
    LAST DEPLOYED: Thu Feb 18 10:05:44 2021
    NAMESPACE: default
    STATUS: deployed
    REVISION: 1
    ```

{{% alert title="Note" color="primary" %}}
When installing K8ssandra on newer versions of Kubernetes (v1.19+), some warnings may be visible on the command line related to deprecated API usage.  This is currently a known issue and will not impact the provisioning of the cluster.

```bash
W0128 11:24:54.792095  27657 warnings.go:70] 
apiextensions.k8s.io/v1beta1 CustomResourceDefinition is 
deprecated in v1.16+, unavailable in v1.22+; 
use apiextensions.k8s.io/v1 CustomResourceDefinition
```

For more information, check out issue [#267](https://github.com/k8ssandra/k8ssandra/issues/267).
{{% /alert %}}

## Verify your K8ssandra installation

> Depending upon your K8 configuration, initialization of your K8ssandra installation can take a few minutes. To check the status of your K8ssandra deployment, use the `kubectl` command:

Reviewer: Replace K8 with K8s

```suggestion
Depending upon your K8s configuration, resources available, and internet speed, initialization of your K8ssandra installation can take a few minutes. To check the status of your K8ssandra deployment, use the `kubectl` command:
```

Reviewer: Do we want to bring up a rough estimate on time?

- ~2 minutes per C* node from image pull complete to boostrapped (sequential operation)
- Stargate does not start until at least 1 C* node is available
- Stargate does not start health checks for 240 seconds

default       k8ssandra-cass-operator-6666588dc5-rrbwx              1/1     Running     0          15m
default       k8ssandra-dc1-default-sts-0                           2/2     Running     0          15m
default       k8ssandra-dc1-stargate-7db4dbfdd5-dxk5v               1/1     Running     0          15m
default       k8ssandra-grafana-b6f7978c4-cwdwg                     2/2     Running     0          15m
default       k8ssandra-kube-prometheus-operator-5556885bd6-2tn4t   1/1     Running     0          15m
default       k8ssandra-reaper-k8ssandra-64cc9d57d9-rc2z4           1/1     Running     0          12m
default       k8ssandra-reaper-k8ssandra-schema-pxdcz               0/1     Completed   0          12m
default       k8ssandra-reaper-operator-cc46fd5f4-hlmck             1/1     Running     0          15m
default       prometheus-k8ssandra-kube-prometheus-prometheus-0     2/2     Running     1          15m
```

NOTE: k8ssandra prefixes the pod names. Stargate takes 4 minutes to register as ready.

Once all the pods are in the `Running` or `Completed` state, you can check the health of your K8ssandra cluster.

{{% alert title="Tip" color="success" %}}

> The pod `k8ssandra-reaper-k8ssandra-schema-xxxxx` runs once and does not persist.

Reviewer: ```suggestion
The pod `k8ssandra-reaper-k8ssandra-schema-xxxxx` is a one-time job that runs to completion. It is visible in K8s to allow for troubleshooting and viewing of logs.
```

{{% /alert %}}

> The actual Cassandra node name from the listing above is `k8ssandra-dc1-default-sts-0`. If you've configured multiple nodes, `sts-0` will increment.

Reviewer: The name is composed of multiple components. We may want to call those out:

- `k8ssandra` - cluster name
- `dc1` - datacenter name
- `default-sts` - rack name + `sts` (meaning stateful set)
- `0` - node id within the rack, note this value is ordinal and generated by the underlying stateful set

We'll use node name `k8ssandra-dc1-default-sts-0` throughout the following sections.

To check the health of your K8ssandra cluster:

1. Confirm that the Cassandra operator is `Ready`:

    ```bash
    kubectl describe CassandraDataCenter dc1 | grep "Cassandra Operator Progress:"
       Cassandra Operator Progress:  Ready
    ```

1. Get K8ssandra superuser credentials:

    * K8ssandra superuser name:

        ```bash

>         kubectl get secret k8ssandra-superuser -o jsonpath="{.data.username}" | base64 --decode | more

Reviewer: I'm not sure we want to include `more` here. The purpose being that there is a highlighted `%` in your shell. This is a [behavior of zshell and specific to your environment](https://unix.stackexchange.com/questions/167582/why-zsh-ends-a-line-with-a-highlighted-percent-symbol). The `more` and `(END)` blocks are a bit distracting in my opinion.

Reviewer: Consider taking this value and placing it in an environment variable that you can then reference throughout the rest of this guide.

```suggestion
        K8SSANDRA_USERNAME=`kubectl get secret k8ssandra-superuser -o jsonpath="{.data.username}" | base64 --decode`
        echo "K8ssandra Username: $K8SSANDRA_USERNAME"
        K8ssandra Username: ZGVtby1zdXBlcnVzZXI=
```

(Note the above has not been tested)

        k8ssandra-superuser
        (END)
        ```

    * K8ssandra superuser password:

        ```bash

>         kubectl get secret k8ssandra-superuser -o jsonpath="{.data.password}" | base64 --decode | more

Reviewer: Same issue as above re: `more`

Reviewer: ```suggestion
        K8SSANDRA_PASSWORD=`kubectl get secret k8ssandra-superuser -o jsonpath="{.data.password}" | base64 --decode`
        echo "K8ssandra Password: $K8SSANDRA_PASSWORD"
        K8ssandra Password: Q19WSDRxdzJRdUFldnczTDEwYmpKRGJzXzM0a1R1RFJOUjJVNmNPdlNKV19sLTk4Z1lsS1VB
```

>         PGo8kROUgAJOa8vhjQrE49Lgruw7s32HCPyVvcfVmmACW8oUhfoO9A

Reviewer: We should highlight that the value returned in their environment will most likely be different and they should not copy this value. @jakerobb and others have suggested taking this value and placing it in an environment variable that you can then reference throughout the rest of this guide.

        (END)
        ```

1. Run `nodetool status`, using the Cassandra node name `k8ssandra-dc1-default-sts-0`, and passing the superuser name and password. Verify that the node is in the state `UN` or Up Normal:

    ```bash

>     kubectl exec -it k8ssandra-dc1-default-sts-0 -c cassandra -- nodetool -u k8ssandra-superuser -pw PGo8kROUgAJOa8vhjQrE49Lgruw7s32HCPyVvcfVmmACW8oUhfoO9A status

Reviewer: Update to use the environment variables we set

```suggestion
    kubectl exec -it k8ssandra-dc1-default-sts-0 -c cassandra -- nodetool -u $K8SSANDRA_USERNAME -pw $K8SSANDRA_PASSWORD status
```

Reviewer: Do we want to use nodetool status here? Should it be moved to the DevOps / SRE persona section?

Reviewer: @jeffreyscarpenter I'm curious about your opinion here.

    Datacenter: dc1
    ===============
    Status=Up/Down
    |/ State=Normal/Leaving/Joining/Moving
    --  Address      Load       Owns    Host ID                               Token                                    Rack
    UN  10.244.1.12  215.3 KiB  ?       75e52e51-edc9-49f8-84f6-f044999ac130  -1080085985719557225                     default

    Note: Non-system keyspaces don't have the same replication settings, effective ownership information is meaningless
    ```

> {{% alert title="Tip" color="success" %}}

Reviewer: With the environment variable changes above this whole alert can go away

Save the superuser name and password for use in future steps.
{{% /alert %}}

## Stopping and starting Cassandra {#cassandra-operations}

Before shutting down your Kubernetes cluster, you'll want to make sure you cleanly shut down your Cassandra datacenters. You can do that using the `kubectl patch` command and setting the `spec:stopped` property to either `true` (stopped) or `false` (running).

### Shut down Cassandra

> To shut down a Cassandra datacenter:

Reviewer: Should we combine the command structure here with the sample below?

To shut down a Cassandra datacenter (while preserving the data) you can patch the CassandraDatacenter resource created by K8ssandra.

```bash
$ kubectl get cassdc
NAME   AGE
dc1    18h

$ kubectl patch cassdc dc1 --type merge -p '{"spec":{"stopped":true}}'
```

Author: You know I'm wondering if this is even worthwhile. It kills the entire pod so there's nothing further you can do. What's the use case here?

```bash
kubectl patch cassdc <datacenter-name> --type merge -p '{"spec":{"stopped":true}}'
```

Example:

```bash
kubectl patch cassdc dc1 --type merge -p '{"spec":{"stopped":true}}'
cassandradatacenter.cassandra.datastax.com/dc1 patched
```

### Start up Cassandra

> To start up a Cassandra datacenter

Reviewer: Starting a stopped Cassandra datacenter is the reverse of stopping one. Setting the `stopped` parameter to `false` will cause all pods to be restarted with their data intact.

```bash
$ kubectl get cassdc
NAME   AGE
dc1    18h

$ kubectl patch cassdc dc1 --type merge -p '{"spec":{"stopped":false}}'
```
